File: sandbox-runner/patterns/saga/typescript/starter.ts
import { Client, Connection } from "@temporalio/client";

import { TASK_QUEUE, WORKFLOW_ID_PREFIX, type TransferDetails } from "./shared";
import { transferMoneyWorkflow } from "./workflows";

async function main(): Promise<void> {
  const connection = await Connection.connect();
  try {
    const client = new Client({ connection });
    const transferId = `${WORKFLOW_ID_PREFIX}-${Date.now()}`;
    const details: TransferDetails = {
      transferId,
      fromAccount: "alice",
      toAccount: "bob",
      amount: 100,
    };

    const handle = await client.workflow.start(transferMoneyWorkflow, {
      args: [details],
      taskQueue: TASK_QUEUE,
      workflowId: transferId,
    });
    console.log(`Started workflow: ${transferId}`);

    const result = await handle.result();
    console.log(result);
    console.log(`Open the Temporal UI and search for '${transferId}' to see the saga history.`);
  } finally {
    await connection.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: sandbox-runner/patterns/saga/go/starter.go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"go.temporal.io/sdk/client"
)

func main() {
	c, err := client.Dial(client.Options{HostPort: "localhost:7233"})
	if err != nil {
		log.Fatalln("Unable to create client:", err)
	}
	defer c.Close()

	transferID := fmt.Sprintf("%s-%d", WorkflowIDPrefix, time.Now().UnixMilli())
	details := TransferDetails{
		TransferID:  transferID,
		FromAccount: "alice",
		ToAccount:   "bob",
		Amount:      100,
	}

	we, err := c.ExecuteWorkflow(
		context.Background(),
		client.StartWorkflowOptions{
			ID:        transferID,
			TaskQueue: TaskQueue,
		},
		TransferMoneyWorkflow,
		details,
	)
	if err != nil {
		log.Fatalln("Unable to execute workflow:", err)
	}
	fmt.Printf("Started workflow: %s\n", we.GetID())

	var result string
	if err := we.Get(context.Background(), &result); err != nil {
		log.Fatalln("Workflow failed:", err)
	}
	fmt.Println(result)
	fmt.Printf("Open the Temporal UI and search for '%s' to see the saga history.\n", transferID)
}
